plugins/log.spec.js

```diff
@@ -0,0 +1,113 @@
+const assert = require('assert');
+const EventEmitter = require('events');
+
+const { Client } = require('..');
+const Encoding = require('./encoding');
+const Log = require('./log');
+
+describe('log plugin', () => {
+    let client;
+    let emitter;
+
+    beforeEach(async() => {
+        emitter = new EventEmitter();
+
+        const logger = (data) => {
+            emitter.emit('log', data);
+        };
+
+        client = await new Client('amqp://guest:guest@127.0.0.1:5672', {
+            plugins: [new Encoding('json'), new Log(logger)]
+        })
+            .start();
+    });
+
+    afterEach(() => client.close());
+
+    describe('publishes', () => {
+        it('it should log messages published to a direct exchange', (done) => {
+            // Ensure logged data is as expected
+            const check = (data) => {
+                assert.deepStrictEqual(data.content, { string: 'string' });
+                assert.strictEqual(data.routingKey, 'a.routing.key');
+                assert.strictEqual(data.exchange, 'amq.direct');
+                assert.strictEqual(data.action, 'publish');
+                done();
+            };
+
+            // Message is logged twice, once when published, once when consumed and acknowledged
+            emitter.once('log', check);
+
+            // Subscribe and acknowledge message
+            client
+                .queue('foo', { durable: false, exclusive: true })
+                .subscribe('a.routing.key', (msg) => msg.ack())
+                .on('error', done);
```

Comment:
@naggingant I ended up leaving these. I do not fully understand why, but removing them causes an error. Maybe the queue needs to be asserted or something before publishing?

Reply:
Quickly checked this and it seems like it's because the tests are done *right after* the call to `publish()` and it triggers the after-each hook, which starts to close the channel also at the same time
Maybe a natural way is to wait for the message to be published (i.e. awaiting `client.publish()`) before calling `done()`

Reply:
Hmm, I see. Do you have any idea why it works when the subscriber is present?

Reply:
This, I don't know exactly, but it seems like it affects the internal state and assumes the publish was made successfully (in the sense of publisher confirmation, I won't go deeper here)
Actually my point here is more on the fact that the connection is destroyed before it is confirmed published, in either way (you can check this by putting some logs for example), so the "right" way kind of is to not call `done()` before `.publish()` resolves
Do I make sense? :sweat_smile:

Reply:
Ah, yes I understood that part thanks. I was just having trouble finding a good way to do it due to the events and asynchronous nature of the test. I revisited to try again though and found a way I'm happy with in 720667d.